src/wallet/rpcwallet.cpp

```diff
@@ -972,6 +990,153 @@ UniValue sendmany(const JSONRPCRequest& request)
+            "\nOptions:\n"
+            "  \"privkeys\":[\"bitcoinprivkey\",...]   (array of strings, required) An array of WIF private key(s)\n"
+            "  \"label\":\"actuallabelname\"           (string, optional) Label for received bitcoins\n"
+            "  \"comment\":\"description\"             (string, optional) Local comment for the receive transaction\n"
+        );
+
+    // NOTE: It isn't safe to sweep-and-send in a single action, since this would leave the send missing from the transaction history
+
+    RPCTypeCheck(request.params, boost::assign::list_of(UniValue::VOBJ));
+
+    // Parse options
+    std::set<CScript> setscriptSearch;
+    CBasicKeyStore tempKeystore;
+    CMutableTransaction tx;
+    std::string strLabel, strComment;
+    CAmount nTotalIn = 0;
+    for (const std::string& optname : request.params[0].getKeys()) {
+        const UniValue& optval = request.params[0][optname];
+        if (optname == "privkeys") {
+            const UniValue& privkeys_a = optval.get_array();
+            for (size_t privkey_i = 0; privkey_i < privkeys_a.size(); ++privkey_i) {
+                const UniValue& privkey_wif = privkeys_a[privkey_i];
+                std::string strSecret = privkey_wif.get_str();
+                CKey key;
+                CPubKey pubkey;
+                ParseWIFPrivKey(strSecret, key, pubkey);
+
+                tempKeystore.AddKey(key);
+                CKeyID vchAddress = pubkey.GetID();
+                CScript script = GetScriptForDestination(vchAddress);
+                if (!script.empty()) {
+                    setscriptSearch.insert(script);
+                }
+                script = GetScriptForRawPubKey(pubkey);
+                if (!script.empty()) {
+                    setscriptSearch.insert(script);
+                }
+            }
+        } else if (optname == "label") {
+            strLabel = AccountFromValue(optval.get_str());
+        } else if (optname == "comment") {
+            strComment = optval.get_str();
+        } else {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Unrecognised option '%s'", optname));
+        }
+    }
+
+    // Ensure keypool is filled if possible
+    {
+        LOCK2(cs_main, pwallet->cs_wallet);
+
+        if (!pwallet->IsLocked()) {
+            pwallet->TopUpKeyPool();
+        }
+    }
+
+    // Reserve the key we will be using
+    CReserveKey reservekey(pwallet);
+    CPubKey vchPubKey;
+    if (!reservekey.GetReservedKey(vchPubKey)) {
+        throw JSONRPCError(RPC_WALLET_KEYPOOL_RAN_OUT, "Error: Keypool ran out, please call keypoolrefill first");
+    }
+
+    // Scan UTXO set for inputs
+    std::vector<CTxOut> vInTXOs;
+    {
+        // Collect all possible inputs
+        std::map<uint256, CCoins> mapcoins;
+        {
+            LOCK(cs_main);
+            mempool.FindScriptPubKey(setscriptSearch, mapcoins);
+            FlushStateToDisk();
+            pcoinsTip->FindScriptPubKey(setscriptSearch, mapcoins);
```

Comment:
It might be appropriate to loop over mapCoins calling CTxMemPool::pruneSpent, on each entry, to avoid trying to sweep from a transaction that is already in the process of being spent.